perf: improve many aspects of DSV performance

triceo opened the pull request:
Less memory allocations.
Less map access.
Cleaner code structure.

Commits are cleanly isolated, so that it is possible to review each idea independently.

Christopher-Chianelli commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/AffectedEntitiesUpdater.java:
```diff
@@ -0,0 +1,199 @@
+        var oldValue = shadowVariableLoopedDescriptor.getValue(affectedEntity);
+        if (!Objects.equals(oldValue, isEntityLooped)) {
+            // TODO what if we don't let users treat this as a shadow var?
+            //  Nobody can hook up to it => we need not trigger events?
```
We need to trigger events for constraint streams; i.e.
```java
cf.forEach(ShadowEntity.class)
  .filter(ShadowEntity::isEntityLooped)
  .penalize(HardSoftScore.ONE_HARD);
```

Christopher-Chianelli reviewed (changes requested):
Changes look good; suggestions within.

triceo commented:
@Christopher-Chianelli I addressed all your comments.
I also added extra commits with more improvements.

Please check the final commit - it seems to have broken a test, although I wasn't able to figure out how. As far as I can tell, the change is functionally equivalent to what was there previously. Please check if you can find out what went wrong - worst case, I revert the final commit.

(But I'd rather not to, because that commit takes MJS from 4.5G heap permanently occupied down to 3G heap.)
